source/server/hot_restart_impl.h

```diff
@@ -10,63 +10,43 @@
 
 #include "envoy/common/platform.h"
 #include "envoy/server/hot_restart.h"
-#include "envoy/server/options.h"
-#include "envoy/stats/stats_options.h"
 
 #include "common/common/assert.h"
-#include "common/stats/raw_stat_data.h"
+#include "common/stats/heap_stat_data.h"
+
+#include "server/hot_restarting_child.h"
+#include "server/hot_restarting_parent.h"
 
 namespace Envoy {
 namespace Server {
 
+// Increment this whenever there is a shared memory / RPC change that will prevent a hot restart
+// from working. Operations code can then cope with this and do a full restart.
+const uint64_t HOT_RESTART_VERSION = 11;
+
 /**
  * Shared memory segment. This structure is laid directly into shared memory and is used amongst
  * all running envoy processes.
  */
-class SharedMemory {
-public:
-  static void configure(uint64_t max_num_stats, uint64_t max_stat_name_len);
-  static std::string version(uint64_t max_num_stats, const Stats::StatsOptions& stats_options);
-
-  // Made public for testing.
-  static const uint64_t VERSION;
-
-  int64_t maxStats() const { return max_stats_; }
-
-private:
-  struct Flags {
-    static const uint64_t INITIALIZING = 0x1;
-  };
-
-  // Due to the flexible-array-length of stats_set_data_, c-style allocation
-  // and initialization are necessary.
-  SharedMemory() = delete;
-  ~SharedMemory() = delete;
-
-  /**
-   * Initialize the shared memory segment, depending on whether we should be the first running
-   * envoy, or a host restarted envoy process.
-   */
-  static SharedMemory& initialize(uint64_t stats_set_size, const Options& options);
-
-  /**
-   * Initialize a pthread mutex for process shared locking.
-   */
-  void initializeMutex(pthread_mutex_t& mutex);
-
+struct SharedMemory {
   uint64_t size_;
   uint64_t version_;
-  uint64_t max_stats_;
-  uint64_t entry_size_;
-  std::atomic<uint64_t> flags_;
   pthread_mutex_t log_lock_;
   pthread_mutex_t access_log_lock_;
-  pthread_mutex_t stat_lock_;
-  pthread_mutex_t init_lock_;
-  alignas(BlockMemoryHashSet<Stats::RawStatData>) uint8_t stats_set_data_[];
-
-  friend class HotRestartImpl;
+  std::atomic<uint64_t> flags_;
 };
+static const uint64_t SHMEM_FLAGS_INITIALIZING = 0x1;
```

Comment:
suggest ShmemFlagInitializing per slight preference in https://github.com/envoyproxy/envoy/blob/master/STYLE.md

>  In the Envoy codebase we use ConstantVar or CONSTANT_VAR. If you pick CONSTANT_VAR, please be certain the name is globally significant to avoid potential conflicts with #defines, which are not namespace-scoped, and may appear in externally controlled header files.

Reply:
This is pre-existing code. These seem like they belong in a followup; this PR is already plenty big.

Reply:
+1 let's fix in a follow up.